[issue #1215] Cannot login to accesspoint on new install

TD-er opened the pull request:
Since event based wifi, it became more apparent the ESP8266 has some issues with running AP and STA mode simultaneous.
It appears this can only work when the channel remains constant. But when trying to connect to an accesspoint that channel will change.
This means Soft AP and STA cannot be active at the same time when searching for an accesspoint to connect to. For example login to the Soft AP to correct some settings or replace an accesspoint would not be possible.

This fix deals with all of these issues and keeps the fast try/fail connection attempts offered by the event based wifi.
In short:
- 3x attempt to connect to AP1 (each attempt takes about 2.5 seconds when failing)
- 3x attempt to connect to AP2
- Enable accesspoint

When accesspoint is active, the connect attempts will be done at a slower pace and STA mode is disabled. STA mode is also disabled as soon as someone connects to the Soft AP.
When the settings are saved, a retry to connect like normal is started almost immediately. N.B. This will result in a disconnect of the connected users to the Soft AP.

So the connection time is:
- AP1 is valid: 4 - 10 seconds
- AP1 is not valid, AP2 is: 12 - 20 seconds
- AP1 and AP2 are not valid: +/- 18 seconds before activating Soft AP

TD-er committed:
[issue #1215] Cannot login to accesspoint on new install

Since event based wifi, it became more apparent the ESP8266 has some issues with running AP and STA mode simultaneous.
It appears this can only work when the channel remains constant. But when trying to connect to an accesspoint that channel will change.
This means Soft AP and STA cannot be active at the same time when searching for an accesspoint to connect to. For example login to the Soft AP to correct some settings or replace an accesspoint would not be possible.

This fix deals with all of these issues and keeps the fast try/fail connection attempts offered by the event based wifi.
In short:
- 3x attempt to connect to AP1 (each attempt takes about 2.5 seconds when failing)
- 3x attempt to connect to AP2
- Enable accesspoint

When accesspoint is active, the connect attempts will be done at a slower pace and STA mode is disabled. STA mode is also disabled as soon as someone connects to the Soft AP.
When the settings are saved, a retry to connect like normal is started almost immediately. N.B. This will result in a disconnect of the connected users to the Soft AP.

So the connection time is:
- AP1 is valid: 4 - 10 seconds
- AP1 is not valid, AP2 is: 12 - 20 seconds
- AP1 and AP2 are not valid: +/- 18 seconds before activating Soft AP

TD-er committed:
[issue #1215] Cannot login to accesspoint on new install

Since event based wifi, it became more apparent the ESP8266 has some issues with running AP and STA mode simultaneous.
It appears this can only work when the channel remains constant. But when trying to connect to an accesspoint that channel will change.
This means Soft AP and STA cannot be active at the same time when searching for an accesspoint to connect to. For example login to the Soft AP to correct some settings or replace an accesspoint would not be possible.

This fix deals with all of these issues and keeps the fast try/fail connection attempts offered by the event based wifi.
In short:
- 3x attempt to connect to AP1 (each attempt takes about 2.5 seconds when failing)
- 3x attempt to connect to AP2
- Enable accesspoint

When accesspoint is active, the connect attempts will be done at a slower pace and STA mode is disabled. STA mode is also disabled as soon as someone connects to the Soft AP.
When the settings are saved, a retry to connect like normal is started almost immediately. N.B. This will result in a disconnect of the connected users to the Soft AP.

So the connection time is:
- AP1 is valid: 4 - 10 seconds
- AP1 is not valid, AP2 is: 12 - 20 seconds
- AP1 and AP2 are not valid: +/- 18 seconds before activating Soft AP

TD-er commented:
If someone could test this PR before I will merge it tomorrow, that would be nice.
I've spent a lot of hours on this one before it was clear to me what was going on here.

The nice thing is by the way, that the ESP32 does seem to work just fine with Soft AP and STA active. Even when the STA is trying to connect over and over again.

TD-er committed:
Merge remote-tracking branch 'origin/bugfix/SoftAP_event_based_wifi' into bugfix/SoftAP_event_based_wifi

Grovkillen commented:
Tested this:

I was never presented with the IP after the unit connected to the "real" Wifi. My (Android) phone sent me to `connectivitycheck.gstatic.com/setup` instead of the `IP/setup`. If I then misspelled the SSID the unit would retry for a while and then after attempt `#6` it would start the AP,  but in between retry the SSID. So during attempt `#7-#10` the unit was only accessible on my phone in between attempts. (I had to refresh multiple times to access the unit). But after `#10` the unit stopped testing the SSID and it was very responsive.

I noticed that each time I refreshed the main page the serial log would say `done.`.

TD-er commented:
This was an existing config, or a new one?
A clean config will not try to connect to an accesspoint, but immediately run in setup mode.
And it will only stop trying to access the given accesspoint when a client is connected to the Soft AP.

Grovkillen commented:
I made sure a reset was done prior to the test... should I have done it on a existing config?

TD-er commented:
Hmm, my edit was finished after you posted :)

Grovkillen commented:
I "answered" on the Slack-message :smiley:

Grovkillen commented:
I guess the "will stop trying" is taking some seconds to be correct since I would say that "many" seconds passed from when I connected externally until it stopped trying.

TD-er commented:
Maybe I should disable the STA mode as soon as someone connects to the SoftAP.
That would speed-up the loading.

There is something very funky going on with the ESP8266 when it is trying to connect to a non-reachable accesspoint with the Soft AP enabled. It keeps sending "disconnect" events even when no connection has been made an nobody is connected to the SoftAP and also no connection attempt was made.
This is both for the 2.3.0 and 2.4.0 core library.

The ESP32 doesn't have that problem.

You could add a `WiFi.enableSTA(false);` to the end of the `processConnectAPmode()` function in `ESPEasyWifi.ino`

TD-er committed:
[Wifi] Use statemachine to move between Wifi states

Still existing problems:
- Credentials are not being saved from web interface and thus connection doesn't work from setup page.
- First time connect to SoftAP takes a lot of attempts, when the wifi was searching for an existing accesspoint. This searching does prevent the SoftAP to form a stable connection to the client.

Maybe the entire wifi stack has to be reset before switching to SoftAP. Now it is near to impossible to recover from incorrect settings or a changes accesspoint.
Still under investigation.

It looks to be a problem with the

TD-er commented:
Changed the entire Wifi connection state into a single state machine function.
Changing between states is partly done, based on events and partly by user interaction.

Still existing problems:
- Credentials are not being saved from web interface and thus connection doesn't work from setup page.
- First time connect to SoftAP takes a lot of attempts, when the wifi was searching for an existing accesspoint. This searching does prevent the SoftAP to form a stable connection to the client.

Maybe the entire wifi stack has to be reset before switching to SoftAP. Now it is near to impossible to recover from incorrect settings or a changed accesspoint.

For new setups and existing config with correct settings it should work just fine.

Still under investigation.

Grovkillen commented:
Two AP settings, none found (configured correctly = online, then power cycled)
* alternate between trying SSID1 and SSID2
* after a total of 10 tries (5 on each SSID) the softAP is launched

One AP, not found (on boot)
* will try once on boot then launch softAP

One AP, not found (online unit saved uncorrect SSID)
* waited for 15 minutes, no reconnect attempt during this time.
* forced a "wifidisconnect" one try then softAP

Haven't tried to move the unit out of coverage (to simulate a AP going down...)

OBSERVE! If you're on Android then clicking the "Log onto network: ESP_Easy_0" in the notification you will be taken to connectivitycheck.gstatic.com/setup instead of 192.168.4.1/setup

Not able to save settings using the setup page (as you already stated)

Grovkillen commented:
And the unit was very responsive when connecting to 192.168.4.1 from all these scenarios.

Grovkillen commented:
Could we have a FORCE softAP included in the code, maybe if we power cycle with a GPIO low or something?

Grovkillen commented:
Or should we build a ESP Easy programmer out of a ESP8266. It will be able to connect to RX/TX of another unit and flash it, program it. We then only need to add a 3.5mm connection somewhere on the receiving units and plug in the programmer there :)

mstjerna commented:
I would recommend to use the Program/Flash GPIO then! Like holding the button down for 5sek would put the AP mode on.
Not during boot but when the ESP is already running... Some other projects use this button to do a factory reset or enter ap-mode when double click or holding the button for 5-10sec...
And the button is already there for the NodeMCU modules and all the common flash programmers :-)

TD-er commented:
There already is a command to start SoftAP.
But the thing is, it just looks like the AP is not detecting client connects when it was scanning for an accesspoint previously.
It appears to be a general problem with the ESP8266 devices, since the channel is shared in hardware and channel hopping does something strange to the SoftAP stability.
I already tried to switch Wifi off and on to AP mode, but that didn't work (but it was already past 3am by then)

mstjerna commented:
And at the same time add the mDNS option to find the ESP on the network once it get IP... This would ease up the boarding process a loot. 
I don't always get the IP-address from the webpage on my mobile. Would ease up for users without knowledge or the skill to find it from the Router.

Grovkillen commented:
Yes and serial command is not something many users use (I think). A button to enable this would be at least the official way of doing it without connecting a computer (or a EASY-programmer ;) )

mstjerna commented:
Have a look at this if you arn't already familiar with the software. They use the button to do:

Button interface

- Click to toggle relays
- Double click to enter AP mode (only main button)
- Long click (>1 second) to reboot device (only main button)
- Extra long click (>10 seconds) to go back to factory settings (only main button)

https://github.com/xoseperez/espurna

TD-er commented:
Tasmota has something similar with the button.
A flash-tool could do the serial commands to set the wifi.

But still I will get it to work, and then adding the mDNS option would be added to I guess.
There is already code for that, but behind a define flag. Not sure why.

TD-er commented:
Oh and deciding a new connect attempt is useless now only takes about 2.5 seconds on my accesspoints. (I look at the disconnect reason)

TD-er committed:
[WiFi] Event based wifi, now AP mode works

It appeared to be related to the core 2.3.0 version. The 2.4.0 does just work. Also changed a number of other things to improve stability.

TD-er committed:
[wifi] Make it work for ESP32 again

TD-er committed:
Merge branch 'mega' into bugfix/SoftAP_event_based_wifi

TD-er committed:
[wifi] Fix wifi AP setup mode for ESP32.

TD-er commented:
With this latest fixed, all seem to work, but I get some strange issues with Core 2.4.0 and saving settings of WiFi on ESP8266.
This results in chunking errors.

Can anyone verify this behavior?

Also the free memory on core 2.4.0 is quite low.

TD-er committed:
[wifi] Event based wifi, improve scan information.

TD-er committed:
Merge remote-tracking branch 'upstream/mega' into bugfix/SoftAP_event_based_wifi

TD-er commented:
For the normal builds, Core 2.4.0 is being used.
For dev and test, core 2.3.0 is used, since they have too little memory left with core 2.4.0

susisstrolch commented:
Just a note about Core 2.4.0...
You can choose (at least in Arduino IDE) between different lwIP variants:

- v2 Lower Memory
- v2 Higher Bandwith
- v1.4 Higher Bandwith

At least the 'v2 Lower Memory' will lead to TCP fragmentation problems (the packages are send out of order) if MTU is exceeded.
'v2 Higher Bandwith' is working fine so far...
